src/compiler/program.rs

```diff
@@ -0,0 +1,169 @@
+        &self,
+        _source_file: &SourceFile,
+    ) -> Vec<Box<dyn Diagnostic>> {
+        vec![]
+    }
+}
+
+impl Program for ProgramConcrete {
+    fn get_semantic_diagnostics(&self) -> Vec<Box<dyn Diagnostic>> {
+        self.get_diagnostics_helper(ProgramConcrete::get_semantic_diagnostics_for_file)
+    }
+
+    fn get_source_files(&self) -> &[Rc<SourceFile>] {
+        &self.files
+    }
+}
+
+struct CreateProgramHelperContext<'a> {
+    processing_other_files: &'a mut Vec<Rc<SourceFile>>,
+    host: &'a dyn CompilerHost,
+    current_directory: &'a str,
+}
+
+pub fn create_program(root_names_or_options: CreateProgramOptions) -> impl Program {
+    let CreateProgramOptions { root_names } = root_names_or_options;
+
+    let mut processing_other_files: Option<Vec<Rc<SourceFile>>> = None;
+    let mut files: Vec<Rc<SourceFile>> = vec![];
+
+    let host = create_compiler_host();
+
+    let current_directory = host.get_current_directory();
+
+    let structure_is_reused = StructureIsReused::Not;
+    if !matches!(structure_is_reused, StructureIsReused::Completely) {
+        processing_other_files = Some(vec![]);
+        let mut processing_other_files_present = processing_other_files.unwrap();
+        let mut helper_context = CreateProgramHelperContext {
+            processing_other_files: &mut processing_other_files_present,
+            host: &host,
+            current_directory: &current_directory,
+        };
+        for_each(root_names, &mut |name, _index| {
+            process_root_file(&mut helper_context, name);
+            Option::<()>::None
+        });
+
+        files = processing_other_files_present;
+        println!("{:?}", files);
+        processing_other_files = None;
+    }
+
+    ProgramConcrete::new(files)
+}
+
+fn process_root_file(helper_context: &mut CreateProgramHelperContext, file_name: &str) {
+    process_source_file(helper_context, &normalize_path(file_name));
+}
+
+fn get_source_file_from_reference_worker(
+    file_name: &str,
+    get_source_file: &mut dyn FnMut(&str) -> Option<Rc<SourceFile>>,
+) -> Option<Rc<SourceFile>> {
+    get_source_file(file_name)
+}
+
+fn process_source_file(helper_context: &mut CreateProgramHelperContext, file_name: &str) {
+    get_source_file_from_reference_worker(file_name, &mut |file_name| {
+        find_source_file(helper_context, file_name)
+    });
+}
+
+fn find_source_file(
+    helper_context: &mut CreateProgramHelperContext,
+    file_name: &str,
+) -> Option<Rc<SourceFile>> {
+    find_source_file_worker(helper_context, file_name)
+}
+
+fn find_source_file_worker(
+    helper_context: &mut CreateProgramHelperContext,
+    file_name: &str,
+) -> Option<Rc<SourceFile>> {
+    let _path = to_path(helper_context, file_name);
+
+    let file = helper_context.host.get_source_file(file_name);
+
+    file.map(|file| {
+        let file = Rc::new(file);
+        helper_context.processing_other_files.push(file.clone());
+        file
```

Comment:
The reason that I made these `SourceFile`'s wrapped in `Rc`'s is that here `findSourceFileWorker()` both appends a newly-created `SourceFile` to `processing_other_files` and returns it. So it seems like I couldn't make `processing_other_files` a `Vec` of _references_ (`Vec<&SourceFile>`) because a reference from here would only live until `find_source_file_worker()` returns, nor a `Vec<Box<SourceFile>>` because a box takes ownership of the boxed value (so then I couldn't return it from `find_source_file_worker()` too), and it didn't seem like a good idea to make `SourceFile`'s `Clone`-able (and so store a separate cloned copy in `processing_other_files`) since they contain entire AST's. So I used `Rc`'s